File: js/src/filesystem.ts
import type { Sandbox } from "./sandbox.js";
import type { Command } from "./command.js";
import type { FileStat, WriteFileParams } from "./types.js";

function assertCommandOk(r: Command, what: string): void {
  if (r.exitCode === 0) return;
  const tail = r.output.trim();
  throw new Error(`${what} failed (exit ${r.exitCode})${tail ? `: ${tail}` : ""}`);
}

export class FileSystem {
  constructor(private readonly sandbox: Sandbox) {}

  async writeFile(path: string, content: string): Promise<void>;
  async writeFile(params: WriteFileParams): Promise<void>;
  async writeFile(pathOrParams: string | WriteFileParams, content?: string): Promise<void> {
    const params =
      typeof pathOrParams === "string"
        ? { path: pathOrParams, content: content ?? "" }
        : pathOrParams;
    return this.sandbox.writeFile(params);
  }

  async exists(path: string): Promise<boolean> {
    const r = await this.sandbox.runCommand("test", ["-e", path]);
    return r.exitCode === 0;
  }

  async stat(path: string): Promise<FileStat> {
    const r = await this.sandbox.runCommand("stat", ["--format=%F|%s|%Y|%a", path]);
    assertCommandOk(r, `stat ${path}`);
    const line = r.output.trim();
    const parts = line.split("|");
    if (parts.length < 4) {
      throw new Error(`unexpected stat output for ${path}: ${line}`);
    }
    const [fileType, sizeStr, mtimeSecStr, mode] = parts;
    return {
      fileType,
      size: Number(sizeStr),
      mtime: new Date(Number(mtimeSecStr) * 1000),
      mode,
    };
  }

  async readdir(path: string): Promise<string[]> {
    const r = await this.sandbox.runCommand("ls", ["-1A", path]);
    assertCommandOk(r, `readdir ${path}`);
    return r.output.split("\n").filter(Boolean);
  }

  async rm(path: string, opts?: { recursive?: boolean }): Promise<void> {
    const args = opts?.recursive ? ["-rf", path] : [path];
    const r = await this.sandbox.runCommand("rm", args);
    assertCommandOk(r, `rm ${path}`);
  }

  async rename(oldPath: string, newPath: string): Promise<void> {
    const r = await this.sandbox.runCommand("mv", [oldPath, newPath]);
    assertCommandOk(r, `rename ${oldPath} -> ${newPath}`);
  }

  async copyFile(src: string, dest: string): Promise<void> {
    const r = await this.sandbox.runCommand("cp", [src, dest]);
    assertCommandOk(r, `copyFile ${src} -> ${dest}`);
  }

  async chmod(path: string, mode: string | number): Promise<void> {
    const m = typeof mode === "number" ? mode.toString(8) : mode;
    const r = await this.sandbox.runCommand("chmod", [m, path]);
    assertCommandOk(r, `chmod ${path}`);
  }

  async symlink(target: string, path: string): Promise<void> {
    const r = await this.sandbox.runCommand("ln", ["-s", target, path]);
    assertCommandOk(r, `symlink ${path}`);
  }
}

File: python/lightning_sdk/sandbox/filesystem.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone

from lightning_sdk.sandbox.base import SandboxInstance, WriteFileParams
from lightning_sdk.sandbox.command import Command


def _assert_command_ok(r: Command, what: str) -> None:
    if r.exit_code == 0:
        return
    tail = r.output.strip()
    msg = f"{what} failed (exit {r.exit_code})"
    if tail:
        msg += f": {tail}"
    raise RuntimeError(msg)


@dataclass(frozen=True)
class FileStat:
    """Metadata from GNU ``stat --format`` (see :meth:`FileSystem.stat`)."""

    file_type: str
    size: int
    mtime: datetime
    mode: str


class FileSystem:
    """Filesystem helpers that run inside the sandbox via shell commands.

    Use :attr:`~lightning_sdk.sandbox.base.SandboxInstance.fs` on a sandbox instance.
    """

    def __init__(self, sandbox: SandboxInstance) -> None:
        self._sandbox = sandbox

    def write_file(
        self,
        path_or_params: str | WriteFileParams,
        content: str | None = None,
    ) -> None:
        """Write file content via the sandbox files API (same as :meth:`SandboxInstance.write_file`)."""
        if isinstance(path_or_params, WriteFileParams):
            self._sandbox.write_file(path_or_params.path, path_or_params.content)
        else:
            self._sandbox.write_file(path_or_params, content or "")

    def exists(self, path: str) -> bool:
        r = self._sandbox.run_command("test", ["-e", path])
        return r.exit_code == 0

    def stat(self, path: str) -> FileStat:
        r = self._sandbox.run_command("stat", ["--format=%F|%s|%Y|%a", path])
        _assert_command_ok(r, f"stat {path}")
        line = r.output.strip()
        parts = line.split("|")
        if len(parts) < 4:
            raise RuntimeError(f"unexpected stat output for {path}: {line}")
        file_type, size_str, mtime_sec_str, mode = parts[:4]
        mtime_sec = int(mtime_sec_str)
        return FileStat(
            file_type=file_type,
            size=int(size_str),
            mtime=datetime.fromtimestamp(mtime_sec, tz=timezone.utc),
            mode=mode,
        )

    def readdir(self, path: str) -> list[str]:
        r = self._sandbox.run_command("ls", ["-1A", path])
        _assert_command_ok(r, f"readdir {path}")
        return [line for line in r.output.split("\n") if line]

    def rm(self, path: str, *, recursive: bool = False) -> None:
        args = ["-rf", path] if recursive else [path]
        r = self._sandbox.run_command("rm", args)
        _assert_command_ok(r, f"rm {path}")

    def mkdir(self, path: str, *, recursive: bool = False) -> None:
        args = ["-p", path] if recursive else [path]
        r = self._sandbox.run_command("mkdir", args)
        _assert_command_ok(r, f"mkdir {path}")

    def rename(self, old_path: str, new_path: str) -> None:
        r = self._sandbox.run_command("mv", [old_path, new_path])
        _assert_command_ok(r, f"rename {old_path} -> {new_path}")

    def copy_file(self, src: str, dest: str) -> None:
        r = self._sandbox.run_command("cp", [src, dest])
        _assert_command_ok(r, f"copy_file {src} -> {dest}")

    def chmod(self, path: str, mode: str | int) -> None:
        m = format(mode, "o") if isinstance(mode, int) else mode
        r = self._sandbox.run_command("chmod", [m, path])
        _assert_command_ok(r, f"chmod {path}")

    def symlink(self, target: str, path: str) -> None:
        r = self._sandbox.run_command("ln", ["-s", target, path])
        _assert_command_ok(r, f"symlink {path}")
